i857 Implement countdown pause time CLICS API PATCH

johnbrvc opened the pull request:
### Description of what the PR does
Fix _CLICS 202306_ contest  information service to correctly handle the `countdown_pause_time` **RELTIME** property.  This will simply clear the scheduled contest start time (setting it to _undefined_), and remember when the contest was paused (in `ContestInformation`).

Fix contest clock displays to show the countdown is paused, if it is.  A tooltip for the clock display label will show how much time was left before the contest was supposed to start when it was paused.

** NOTE: Must be merged AFTER #864 since this is a branch from that branch.

### Issue which the PR addresses
Fixes #857 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11, Ubuntu 24.04
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
1) Load up a new contest that has a start time defined that is in the future (implies starting the server).  You can use the `samps/clics_sumithello`, and then set a start time in the future after loading up the contest using the "**Times**" tab on the server.
2) Start an _event feeder_ client, and press **Start** so it is listening for connections
3) Start an _administrator_ client - note the countdown clock in the upper left is counting down to the start of the contest
<img width="278" alt="image" src="https://github.com/user-attachments/assets/c9a8b8ca-6f0c-47b2-bf6f-3fa48f4fb025" />

From this point, you have 2 choices of how to test, the easy way by using "curl" and the "practical" way by using a CDS.  Both methods will be described below:

4a) Curl: To pause contest **SumH** and have it tell the CCS it was paused with 2 hours remaining:
```
curl -X PATCH -k \
            https://admin:admin@localhost:50443/contests/SumH \
            -H 'Content-Type: application/json' \
            -d '{"id":"SumH", "start_time":null, "countdown_pause_time":"2:00:00" }'
```
When you pause, the countdown clock will stop and say **Paused** or **Stopped** depending on the client.  Hovering over the **Paused** or **Stop** message will tell you how much time was left when it was paused.  To resume contest **SumH** and have it start at the specified time (2025-04-20T14:00:00-04:00):
```
curl -X PATCH -k \
        https://admin:admin@localhost:50443/contests/SumH \
        -H 'Content-Type: application/json' \
        -d '{"id":"SumH", "start_time":"2025-04-17T14:00:00-04:00", "countdown_pause_time":null }'
```

4b) CDS:  The remaining steps are all to test using a CDS.  Configure an _ICPC Tools_ CDS (at least version 2.6.1177).  For the `cdsConfig.xml `PC2 credentials you can use "**adminsitrator1**" as the user and "**administrator1**" as the password, assuming that is how you have that account configured in PC2.  Also, you'll have to add an "**administrator1**" admin account to the accounts.yaml.  [Here is a TXT file that you can rename as `accounts.yaml`]([accounts.yaml.txt](https://github.com/user-attachments/files/19805971/accounts.yaml.txt)
) and put it in the `config `folder.  The user/password when you log in to the browser will be **administrator1**/**administrator1**.

5b) Start the CDS (From a command prompt: `bin\server run cds`)- it takes awhile - wait until the "_smarter planet_" message comes up in the console window.

```
> [AUDIT   ] CWWKZ0001I: Application CDS started in 1.612 seconds.
> [AUDIT   ] CWWKF0012I: The server installed the following features: [appSecurity-5.0, cdi-4.0, distributedMap-1.0, expressionLanguage-5.0, jndi-1.0, jsonp-2.1, pages-3.1, servlet-6.0, ssl-1.0, transportSecurity-1.0, websocket-2.1].
> [AUDIT   ] CWWKF0011I: The cds server is ready to run a smarter planet. The cds server started in 19.358 seconds.
> 
```

6b) Open a browser to https://localhost:8443/  ( found it works faster if you use the IP address of your PC instead of "localhost", but that's on my machine.
7b) Login (upper right) using:  administrator1/administrator1  (same as in the accounts.yaml and the same as PC2 has for administrator1).  Give it a chance to log in.  Here is the CDS console log:
```
Reading feed cache
Unknown property ignored: state/paused
Unknown property ignored: state/resumed
Unknown property ignored: state/contest_time
Done reading feed cache
```

8b) Select the **Countdown Control** selection in the left pane
9b) You should see the contest countdown clock in the browser window counting down, and it should match the admin countdown.
<img width="836" alt="image" src="https://github.com/user-attachments/assets/a6cc1bbf-34a6-4a64-a105-c1a56fc31add" />

10) Press the: Pause button.  Paused message should appear on the administrator window as well as the CDS browser window.
<img width="299" alt="image" src="https://github.com/user-attachments/assets/10ca4c88-cd7f-4ee8-b36c-71e51974bcf3" />
<img width="824" alt="image" src="https://github.com/user-attachments/assets/f05dedec-0260-415c-be3c-993ada01ceaa" />

11b) Press the: Resume button.  The contest should resume counting down on both the administrator as well as the CDS browser window.  Here is the CDS console log:

```
Start time command: pause
Setting start time to paused at: 23:32:40.000
Adding client-side event to feed cache
Setting contest time at https://localhost:50443/contests/SumH
Start time command: resume
Setting start time to: 2025-04-21T20:15:07.000-04:00 (8:15.07 PM April 21, 2025)
Adding client-side event to feed cache
Setting contest time at https://localhost:50443/contests/SumH
Start time command: pause
Setting start time to paused at: 23:32:39.000
Adding client-side event to feed cache
Setting contest time at https://localhost:50443/contests/SumH
Finished scanning SumH for changed resources in 3ms
```
12b) repeat the process if desired.  You can "set" various countdown resume times using the CDS browser window.

johnbrvc committed:
i_857 Implement countdown pause time patch

Fix CLICS 202306 contest service to correctly handle the countdown_pause_time property.  This will simply clear the schedule start time (setting it to undefined), and remember when the contest was paused (in ContestInformation).
Fix contest clock displays to show the countdown is paused if it is.  Tooltip will show the time the contest was paused.

SamanwaySadhu commented on src/edu/csus/ecs/pc2/ui/ContestClockDisplay.java:
```diff
@@ -551,11 +585,18 @@ public String getScheduleOrRemainingTime(ContestTime contestTime) {
 
+    // This is used for unit tests only at the moment.
+    public String getScheduleOrRemainingTime(ContestTime contestTime) {
+        return(getScheduleOrRemainingTimeAndHint(contestTime)[0]);
```
Put a space after the return

johnbrvc replied:
Fixed in next push.  Actually, I just removed the parens to be consistent with other return values.

johnbrvc committed:
i_857 Fix code style for return

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/ContestService.java:
```diff
@@ -238,7 +237,7 @@ public Response setContestTimes(@Context HttpServletRequest servletRequest, @Con
      * @param sawCountdownPauseTime
      * @return web response
      */
-    private Response HandleContestStartTime(SecurityContext sc, String contestId, String startTimeValueString, boolean sawCountdownPauseTime) {
```
Given that this parameter was removed from the method (a good change), it should also be removed from the JavaDoc (above).

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/ContestService.java:
```diff
@@ -60,6 +60,7 @@
  * {
  *  "id":"wf2016",
  *  "start_time":null
+ *  "countdown_pause_time":"2014-06-25T09:00:00+01"
```
It's nice that this was added to the example.  However, it's a bit strange to have an example that lists the contest id as "wf2016" when the "countdown pause time" is two years prior :)   (Suggest changing the year in the example to "2016" (or else change the id to "2014")  -- even though I recognize it's all a comment and it doesn't REALLY matter.  But *I* noticed it, so somebody else might, and might be confused... )

johnbrvc replied:
I noticed this when I did added it, but figured, eh, nobody would notice (or care).  Anyway, fixed in next push

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/ContestService.java:
```diff
@@ -476,9 +481,12 @@ private boolean setScheduledStart(GregorianCalendar theDate, boolean unPauseCoun
             }
-            // TODO unpause countdown - this will be a flag in "ci"
             // tell the Controller to update the ContestInformation
+            // note: setting a new contest start time (eg theData != null) will automatically start the countdown
```
This comment references a variable named `theData`.  I think it was intended to reference `theDate`.

johnbrvc replied:
Fixed in next push

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/ContestService.java:
```diff
@@ -463,10 +468,10 @@ private GregorianCalendar getDate(String contestId, String startTimeValueString)
      * @return true if the method was successful in setting the scheduled start time; false otherwise
      */
-    private boolean setScheduledStart(GregorianCalendar theDate, boolean unPauseCountdown) {
+    private boolean setScheduledStart(GregorianCalendar theDate, String contestPauseRelTime) {
```
It would be nice if the JavaDoc at the top of this method provided a little bit of insight into how the method handles the fact that it can be called with one parameter null and the other not, OR called with both parameters being null.  (What little explanation of the various modes of operation existed in the documentation is lost with the deletion of the `@param unPauseCountdown` parameter.)

johnbrvc replied:
Fixed in next push

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/ContestService.java:
```diff
@@ -389,9 +388,15 @@ private Response HandleContestCountdownPauseTime(SecurityContext sc, String cont
+            // tell PC2 to stop countdown when clock is 'pauseTime' ms away from start
+            boolean success = setScheduledStart(null, countdownPauseTime);
+            if (success) {
+                controller.getLog().log(Log.INFO, LOG_PREFIX + contestId + ": countdown paused (set contest start time to null)");
```
I think this log message is a litle confusing: it says that the countdown is paused, and that the start time was set to null.  Those are both true -- but it would be clearer if it _also_ included the (formatted) "countdownPauseTime" -- so that the person reading the log understands that _setting the pause time to that value_ is _**also**_ a part of what has happened.

johnbrvc replied:
Fixed in next push

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/ContestService.java:
```diff
@@ -389,9 +388,15 @@ private Response HandleContestCountdownPauseTime(SecurityContext sc, String cont
+            boolean success = setScheduledStart(null, countdownPauseTime);
+            if (success) {
+                controller.getLog().log(Log.INFO, LOG_PREFIX + contestId + ": countdown paused (set contest start time to null)");
+                return Response.ok().entity("Contest countdown paused - start time updated to \"null\" (no scheduled start)").build();
```
Analogous to the above comment: I think it would helpful if the Response message included the "pause time" -- i.e. if the message was (something like) `Contest countdown paused at 00:05:00 and start time updated to "null" (no scheduled start)`.

johnbrvc replied:
Fixed in next push

clevengr reviewed (commented):
Reviewed `src/edu/csus/ecs/pc2/clics/API202306/ContestService.java`.  Left a number of suggestions, none of them showstoppers.

clevengr commented on src/edu/csus/ecs/pc2/core/model/ContestInformation.java:
```diff
@@ -173,6 +173,9 @@ public enum TeamDisplayMask {
 
     private boolean autoStopContest = false ;
 
+    // RELTIME (HH:MM:SS) of when contest countdown was paused (Before start)
```
Actually, the CLICS spec defines RELTIME to be more complex than just `HH:MM:SS`; in particular, it can have more than two _HOUR_ digits (plus some other additional stuff).  I don't think the other stuff is important, but it might be helpful here to point out that more than 24 hours (indeed, more than 99 hours) is allowed.  Someone reading the code and not intimately familiar with the spec might benefit from having this pointed out here.

There are several other places below where `HH:MM:SS` appears; I don't think it's necessary to point it out in all those -- but it might be helpful to point it out once, here...

johnbrvc replied:
Fixed in next push.

clevengr reviewed (commented):
Reviewed `src/edu/csus/ecs/pc2/core/model/ContestInformation.java`.  Left one minor comment; no showstoppers.

clevengr commented on src/edu/csus/ecs/pc2/ui/ContestClockDisplay.java:
```diff
@@ -48,6 +48,14 @@ public class ContestClockDisplay implements ActionListener, UIPlugin {
     private static final long serialVersionUID = 8137635697344335832L;
 
     private static final String NO_SCHEDULE_START_TIME = "No scheduled start";
+    private static final String NO_SCHEDULE_START_HINT = "Still no scheduled start";
```
I feel like the word `SCHEDULE` in these constants should be `SCHEDULED`.
(And yes, I'm aware that one of them was pre-existing...)

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/ui/ContestClockDisplay.java:
```diff
@@ -512,37 +535,48 @@ protected void updateScheduledStartLabels() {
-            
-            /**
-             * Show Remaining time count down. 
+    public String [] getScheduleOrRemainingTimeAndHint(ContestTime contestTime) {
```
I feel like the name of this method should contain the word `Scheduled` (instead of `Schedule`).  As named, it sounds like it's getting either "a schedule [of something] or a Remaining time and hint" -- whereas it's intended to convey that it is getting either "a Scheduled _Time_ or a Remaining Time"...
 (And yes, I'm aware it was pre-existing.  )

Also, it would be nice if there was some minimal JavaDoc explaining what the method is doing and how the parameter is used, especially since it's returning different kinds of things depending on the state of the contest.
(And again, yes I'm aware this method, without any JavaDoc, was pre-existing..)

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/ui/ContestClockDisplay.java:
```diff
@@ -725,17 +775,24 @@ public void setClientFrame(JFrame clientFrame) {
-    
+
+    /**
+     * Set the contest schedule start time from the supplied contest information
```
I think the word `schedule` in this comment should be `scheduled`...

clevengr reviewed (commented):
Reviewed `src/edu/csus/ecs/pc2/ui/ContestClockDisplay.java`.  Left some comments; no showstoppers.

clevengr reviewed (commented):
I've reviewed all four changed files in the PR; left comments but no showstoppers.   I'm moving on to testing the PR.

johnbrvc committed:
i_857 PR Change requests

No functional changes.
Additional comments added.
Incorrect comments fixed.
Some log messages have more detail.
Refactor some member variable names.

clevengr commented:
@johnbrvc I have tried to test the PR following the instructions given in the "Precise steps for testing the PR".  I am able to start a CDS (latest version:  `Prerelease: [CDS v2.6.1191](https://github.com/icpctools/icpctools/releases/download/v2.6.1191/wlp.CDS-2.6.1191.zip)`.  The CDS successfully connects to the PC2 Event Feeder and displays the data for my contest (`samps/sumithello`). 
I have a Scheduled Start time well in the future; clicking on the CDS's Countdown Control item shows a countdown which is indeed counting and does indeed match what's shown on the PC2 Admin screen.

However, when I execute Step 10 (`Press the: Pause button`) (I assume it's talking about about the Pause button on the CDS display), it does not pause the contest (that is, both the CDS countdown and the PC2 Admin countdown continue to decrement.  The CDS displays a small error message next to its countdown clock:  `Unknown error (401)`.  

I've looked in the CDS logs; there's no error listed there.  I'm stuck, at a loss to know what to do.  I _suspect_ this might be an issue in the CDS -- but it also might be an issue with the StartTime endpoint.  I don't have any idea how to work around it so I can complete testing.

Any suggestions?

(Hmmm... maybe I need to restart everything and run the Event Feed client inside a debugger?)

johnbrvc commented:
> @johnbrvc I have tried to test the PR following the instructions given in the "Precise steps for testing the PR". I am able to start a CDS (latest version: `Prerelease: [CDS v2.6.1191](https://github.com/icpctools/icpctools/releases/download/v2.6.1191/wlp.CDS-2.6.1191.zip)`. The CDS successfully connects to the PC2 Event Feeder and displays the data for my contest (`samps/sumithello`). I have a Scheduled Start time well in the future; clicking on the CDS's Countdown Control item shows a countdown which is indeed counting and does indeed match what's shown on the PC2 Admin screen.
> 
> However, when I execute Step 10 (`Press the: Pause button`) (I assume it's talking about about the Pause button on the CDS display), it does not pause the contest (that is, both the CDS countdown and the PC2 Admin countdown continue to decrement. The CDS displays a small error message next to its countdown clock: `Unknown error (401)`.
> 
> I've looked in the CDS logs; there's no error listed there. I'm stuck, at a loss to know what to do. I _suspect_ this might be an issue in the CDS -- but it also might be an issue with the StartTime endpoint. I don't have any idea how to work around it so I can complete testing.
> 
> Any suggestions?
> 
> (Hmmm... maybe I need to restart everything and run the Event Feed client inside a debugger?)

I have seen this from time to time, and so has @SamanwaySadhu .  It appears to be an internal web call the cds makes to itself that is failing.  If you bring up the chrome debugger, "**Network**" tab, you'll see the "pause" or "resume" fails with 401 trying to reach https://localhost:8443/contests/SumH/admin/time/resume  (localhost:8433 is itself).

I asked Tim about it.  We'll see what he says.  As I said, I usually just restart the CDS and let it sit there awhile before attempting to connect to it.  You can also remove the CDS cache in `\users\YOURNAME\AppData\local\Temp\org.icpc.tools.cache.*`

SamanwaySadhu reviewed (approved):
LGTM!
